properly classify shebang comments in the IDE

brettfo opened the pull request:
Building off of PR #4470, this ensures the file-leading shebang comment is classified as trivia so it gets colorized in the IDE.

Tagging @Pilchie @dpoeschl @jasonmalinowski @rchande @balajikris @CyrusNajmabadi @jmarolf @davkean @KevinH-MS for review.

brettfo commented:
And I forgot to tag @DustinCampbell, how embarrassing.

DustinCampbell commented on src/Workspaces/CSharp/Portable/Extensions/SyntaxTriviaExtensions.cs:
```diff
@@ -50,6 +50,11 @@ public static bool IsMultiLineComment(this SyntaxTrivia trivia)
 
+        public static bool IsShebangComment(this SyntaxTrivia trivia)
+        {
+            return trivia.Kind() == SyntaxKind.ShebangCommentTrivia;
```
Based on our API naming conventions, this will should be HashExclamationTrivia. Who from the scripting team is on the hook to follow the API naming guidelines? @tmat? @amcasey? @cston?

tmat replied:
I don't see why it would need to be HashExclatationTrivia. 

SingleLineCommentTrivia is also not called SlashSlashTrivia

DustinCampbell replied:
I think it's probably up to API czars. For XML doc comments, we have things like LessThanExclamationMinusMinus. However, this isn't a new token, so i't probably OK. @AnthonyDGreen?

DustinCampbell commented:
:+1:

Also, since you're updating these existing extension methods, it might be worth adding tests to ensure that other features using them work properly inside shebang comments. For example, does this change keep symbol completion from showing up in shebang comments?

KevinH-MS commented:
:+1:

CyrusNajmabadi commented:
FWIW, TypeScript calls this "ShebangTrivia".

brettfo commented:
@DustinCampbell thanks for the pointer, indeed symbol completion (and others) were triggering inside the shebang.  I've updated the PR that fixes this and adds a test to ensure it doesn't regress.  I've also audited usage of `SyntaxKind.SingleLineCommentTrivia` and updated to handle `ShebangCommentTrivia` as appropriate.

AnthonyDGreen commented:
What is a Shebang comment?

KevinH-MS commented:
@AnthonyDGreen You put it at the start of your .csx or .vbx file so that you can exec your scripts like a binary executable on *nix.

https://en.wikipedia.org/wiki/Shebang_(Unix)

AnthonyDGreen commented:
Given that the shebang terminology is well understood by the community (even having a Wikipedia page), I think the name is fair game.

brettfo commented:
@jasonmalinowski: Updated doc comments as requested.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Syntax/SyntaxNodeRemover.cs:
```diff
@@ -110,10 +110,16 @@ private void AddEndOfLine()
+            /// <summary>
+            /// Returns whether the specified <see cref="SyntaxTrivia"/> token is also the end of the line.  This will
+            /// be true for <see cref="SyntaxKind.EndOfLineTrivia"/>, <see cref="SyntaxKind.SingleLineCommentTrivia"/>,
+            /// <see cref="SyntaxKind.ShebangCommentTrivia"/>, and all precompiler directives.
```
precompiler?

brettfo replied:
Should be preprocessor, good catch :)

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/SyntaxTriviaExtensions.cs:
```diff
@@ -32,12 +32,12 @@ public static bool MatchesKind(this SyntaxTrivia trivia, params SyntaxKind[] kin
         public static bool IsRegularComment(this SyntaxTrivia trivia)
         {
-            return trivia.IsSingleLineComment() || trivia.IsMultiLineComment();
+            return trivia.IsSingleLineComment() || trivia.IsMultiLineComment() || trivia.IsShebangComment();
```
I wouldn't call it a ShebangComment.  It's not a comment.  It's just "IsShebang"

brettfo replied:
I called it `IsShebangComment` to keep it in line with `SyntaxKind.ShebangCommentTrivia` that @KevinH-MS added in an earlier PR.

CyrusNajmabadi replied:
ShebangCommentTrivia seems like the wrong name.  It should be ShebangTrivia.

brettfo replied:
I'm going to deflect this naming question to @AnthonyDGreen and @KevinH-MS.  See Cyrus' comment above.

KevinH-MS replied:
It's treated just like a single-line comment everywhere from a (compiler/language perspective), hence the name.  Originally, it didn't even have it's own ```SyntaxKind```, but I added that after it was suggested that hosts might want to easily distinguish these without comparing the leading ```char```.

CyrusNajmabadi commented:
> It's treated just like a single-line comment everywhere 

That sounds like it's treated like *trivia* then :)

'Comment' has some pretty well defined meaning, and this does not appear to be like a comment to me.

DustinCampbell commented:
@CyrusNajmabadi is right. Looking at this closer and seeing that we're doing things like semantic classification seems a bit weird for a "comment". On the assumption that this is a comment I suggested turning off completion, but that's clearly wrong since the tests show "System.Console.WriteLine" semantically colorized. Why wouldn't we want completion to work on that?

This is definitely trivia, but it's not a comment. The examples in tests seem to indicate that it's more like a custom preprocessor directive of some sort.

KevinH-MS commented:
We should ***not*** be doing semantic classification in a shebang.  It's an entirely syntactic construct.  I agree that it looks/feels like it could be a "ShebangDirectiveTrivia", but we decided that it would not have structure/semantics, so I don't think that's as appropriate.  The intent is that it looks and acts just like a comment, and I think that having "Comment" in the name of the drives home that point.  That said, if compiler/API czars feel otherwise (or if we think it should behave differently...for instance, like disabled text), then we could certainly change it.

jasonmalinowski commented:
@KevinH-MS: it doesn't have structure? It definitely has the line to the command interpreter itself. You could even imagine the compiler would assert some reasonable-well-formedness rules to that (it must start with a leading /, for example.)

KevinH-MS commented:
@jasonmalinowski No, currently, we only enforce the location.  After that, it's just a comment.  It's intended to be generally extensible (free-form), and the host can process it however it wants.

brettfo commented:
I've updated the PR with `ShebangCommentTrivia` being renamed to `ShebangTrivia` (and fixed some merge conflicts) so it can be looked at in context.

While I agree with @jasonmalinowski that it does have structure, the [Wikipedia page](https://en.wikipedia.org/wiki/Shebang_(Unix)) even says that the actual structure of the text following `#!` isn't rigidly adhered to; e.g., should there be a space between `!` and `/`?  Because of that I don't think it should be classified as any sort of directive.  As to @KevinH-MS's comment about it behaving like a single line comment, I agree, just that it has a special place, but I don't like the word 'comment' in there because it sounds like I can do this: `int x = 1; #! set x equal to 1`.

KevinH-MS commented:
@brettfo I would point out that doc comments are "Documentation**Comment**Trivia", but they are only allowed in certain places (or at least they're processed differently if it's not in the *right* place).

That said, I'm not vehemently opposed to the change.
